hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala

```diff
@@ -2465,88 +2465,91 @@ case class MatrixExplodeCols(child: MatrixIR, path: IndexedSeq[String]) extends
+
+  private val entriesFieldType = child.typ.rowType.fieldType(entriesFieldName)
+
+  private val Field(_,
+    TArray(colType@TStruct(_, _), _),
+    colsFieldIdx
+  ) = child.typ.globalType.field(colsFieldName)
+
+  private val newRowType = child.typ.rowType.rename(m)
+
+  entriesFieldType match {
+    case TArray(TStruct(_, _), _) =>
+    case _ => fatal(s"expected entry field to be an array of structs, found ${ entriesFieldType }")
   }
 
   val typ: MatrixType = MatrixType(
-    rowsEntries.typ.globalType,
-    cols.typ.key,
-    cols.typ.rowType,
-    rowsEntries.typ.key,
+    child.typ.globalType.deleteKey(colsFieldName, colsFieldIdx),
+    colKey,
+    colType,
+    child.typ.key,
     newRowType)
 
-  def children: IndexedSeq[BaseIR] = Array(rowsEntries, cols)
+  def children: IndexedSeq[BaseIR] = Array(child)
 
-  def copy(newChildren: IndexedSeq[BaseIR]): UnlocalizeEntries = {
-    assert(newChildren.length == 2)
-    UnlocalizeEntries(
+  def copy(newChildren: IndexedSeq[BaseIR]): CastTableToMatrix = {
+    assert(newChildren.length == 1)
+    CastTableToMatrix(
       newChildren(0).asInstanceOf[TableIR],
-      newChildren(1).asInstanceOf[TableIR],
-      entryFieldName
-    )
+      entriesFieldName,
+      colsFieldName,
+      colKey)
   }
 
-  override def partitionCounts: Option[IndexedSeq[Long]] = rowsEntries.partitionCounts
+  override def partitionCounts: Option[IndexedSeq[Long]] = child.partitionCounts
 
   def execute(hc: HailContext): MatrixValue = {
-    val rowtab = rowsEntries.execute(hc)
-    val coltab = cols.execute(hc)
+    val prev = child.execute(hc)
 
-    val localColType = coltab.typ.rowType.physicalType
-    val localColData: Array[Annotation] = coltab.rvd.mapPartitions { it =>
-      it.map { rv => SafeRow(localColType, rv.region, rv.offset) : Annotation }
-    }.collect
+    val localColData = prev.globals.value.getAs[IndexedSeq[Annotation]](colsFieldIdx)
+    val nCols = localColData.size
 
-    val field = GetField(Ref("row", rowtab.typ.rowType), entryFieldName)
-    val (_, lenF) = ir.Compile[Long, Int]("row", rowtab.typ.rowType,
+    val field = GetField(Ref("row", prev.typ.rowType), entriesFieldName)
+    val (_, lenF) = ir.Compile[Long, Int]("row", child.typ.rowType,
       ir.ArrayLen(field))
 
-    val (_, missingF) = ir.Compile[Long, Boolean]("row", rowsEntries.typ.rowType,
+    val (_, missingF) = ir.Compile[Long, Boolean]("row", child.typ.rowType,
       ir.IsNA(field))
 
-    var rowRVD = rowtab.rvd
-    rowRVD = rowRVD.mapPartitionsWithIndex(rowRVD.typ) { (i, it) =>
+    val rowRVD = prev.rvd.mapPartitionsWithIndex(prev.rvd.typ) { (i, it) =>
       val missing = missingF(i)
       val len = lenF(i)
       it.map { rv =>
         if (missing(rv.region, rv.offset, false)) {
-          fatal("missing entry array value in argument to UnlocalizeEntries")
+          fatal("missing entry array value in argument to CastTableToMatrix")
```

Comment:
This is not incredibly relevant to the scope of this PR, and you don't need to change it here, but these checks could all just be compiled into a single IR function with `Die` nodes.
```
If(IsNA(field),
   Die("missing entry..."),
   If(ApplyComparisonOp(EQ(), ArrayLen(field), nCols),
      Die("wrong length..."),
      Ref("row")))
```